lib/platform/platform.c

```diff
@@ -429,3 +431,174 @@ ssize_t write_from_gas(struct switchtec_dev *dev, int fd,
 {
 	return dev->ops->write_from_gas(dev, fd, src, n);
 }
+
+/**
+ * @brief Copy data to either the local GAS or the remote GAS
+ * @param[in]  dev	Switchtec device handle
+ * @param[out] dest     Destination gas address
+ * @param[in]  src      Source data buffer
+ * @param[in]  n        Number of bytes to transfer
+ */
+void fab_memcpy_to_gas(struct switchtec_dev *dev, void __gas *dest,
+		       const void *src, size_t n)
```

Comment:
> 
> 
> If the point of this is to support remote gas access then shouldn't this have some way of addressing the remote partition ID?
> 
> I don't see any value in this otherwise.

Currently, the firmware is using the GAS of the same interface/partition on remote PAX to serve the access request, e.g., the remote GAS access through the TWI interface will be served by the remote GAS of TWI interface. It was discuessed with firwmare team, they think it's confusing. Most of the GAS fields are actually shared among different interfaces/partitions, but some of them are interface/partition specific. So for most of the GAS locations, there should be no problem. But yes, it's an issue.

Reply:
> 
> 
> Also, this seems pretty scary in that now any host can start changing any registers out from under the kernel. This is a large permission problem that should probably be addressed some how....

What will you suggest? I see most APIs that need GAS access is for event features and I think we should limit the event feature to the local PAX only.

Reply:
I think we should leave it the way it is. Then have another API that allows for remote access. 

It might make sense to add a check in the kernel to ensure the user is root before allowing MRPC commands that provide access to any partitions' GAS as this is dangerous and stepping on the toes of the kernel's responsibility.

Reply:
Could you explain more on the current GAS functions use policy regarding the privilidge?

1. Why do we want to limit the GAS functions to the platform code？
2. Why do we not allow GAS functions be used on the Linux platform?

```
#ifdef SWITCHTEC_LIB_CORE
#error "You should not be using GAS access functions in the core library."
#endif

#ifdef SWITCHTEC_LIB_LINUX
#error "GAS Access functions should not be used on the Linux platform " \
        "as they require full root access."
#endif
```

Reply:
> 
> 
> I think we should leave it the way it is. Then have another API that allows for remote access.
> 
> It might make sense to add a check in the kernel to ensure the user is root before allowing MRPC commands that provide access to any partitions' GAS as this is dangerous and stepping on the toes of the kernel's responsibility.

Because we also have interfaces like ethernet on PAX, so I think it might be too heavy to involve the driver to ensure the user is root. We might leverage getuid()/setuid() to ensure it's a root user that accessing remote GAS.

Reply:
1. Because they require a different privilege set on Linux (they really should also on Windows, but that's the way I things went) and thus shouldn't be used for any operation that might .

2. Because the kernel owns the register space and userspace can really break things if they try to change things (or, as we've seen, they can even attack the system simply by reading because that can swamp the firmware). That's why the gas functions only work as proper root while every other function only requires access to the /dev/switchtec device.

getuid() would not provide any security protection as all an attacker would have to do is rebuild the library with that disabled. setuid() should absolutely not be used by this code.

Reply:
All these make sense. But it seems currently only Linux platform enforced this rule. On other platform like UART, if I grant the access of /dev/ttyUSBx to unprivileged user, gas access on this interface can be performed without root privilege.
```
#ll /dev/ttyUSB2
crw-rw-rw- 1 root dialout 188, 2 May 27 17:09 /dev/ttyUSB2
#./switchtec gas read /dev/ttyUSB2 -a 0x2000
002000 - 0x00008546
```

Because there're special MRPC commands to do GAS READ/WRITE, we might need to pump up the privilige required for switchtec API functions that issuing these GAS access MRPC commands (e.g. through map() and unmap() the /dev/mem, before any GAS access. But it would be an problem if /dev/mem does not exist. Any alternative to suggest?).

Reply:
For UART and I2C, it doesn't really matter as the kernel isn't really involved. Though the user can still shoot themselves in the foot if they have one machine modifying hw registers of another machine using a UART or I2C device. Pretty nasty stuff.

To enforce the permissions of MRPC command just have the kernel detect the specific command and fail if  "capable(CAP_SYS_ADMIN)" is not true. No need to detect if the process was able to map anything, that's nasty.

Reply:
Sorry for the delay.

I'm still a little confused on this policy. For UART or I2C device, when granted access to the device node, the user has access to both MRPC commands and GAS (local/remote). But for the switchtec device node,  the access to the device node only allows for MRPC commands and extra capabililty is needed to allow GAS (remote). I don't think the inband interface is more important then the others, and it's likely that TWI is more widely used.

How about we add the detect of remote (or local and remote) gas access and requiring of CAP_SYS_ADMIN capability in the user library, so any GAS (local/remote) access would either requires root privilege or CAP_SYS_ADMIN capability, regardless of the interface. You may think attacher can remove this limitation and recompile the code, but the attacker still need to get the privilege to access the device node.  I think we are not making it worse but better.

Reply:
It's not that inband is more "important" it's that the kernel owns access to the GAS and can't have userspace interfering. Especially when a DOS attack using the GAS can cause the CPU to stall because our IO is slow.

You can't meaningfully check for permissions in the library. That's nonsense. All an attacker has to do to get by that is to rebuild the library without the check. The check *must* be in the kernel.